python/ray/rllib/utils/sampler.py

```diff
@@ -155,67 +164,146 @@ def _env_runner(env, policy, num_local_steps, horizon, obs_filter, pack):
+        ready_obs = []
+        ready_rnn_states = []
+
+        # Process and record the new observations
+        for eid, raw_obs in unfiltered_obs.items():
+            episode = episodes[eid]
+            filtered_obs = obs_filter(raw_obs)
+            ready_eids.append(eid)
+            ready_obs.append(filtered_obs)
+            ready_rnn_states.append(episode.rnn_state)
+
+            if episode.last_observation is None:
+                episode.last_observation = filtered_obs
+                continue  # This is the initial observation after a reset
+
+            episode.length += 1
+            episode.total_reward += rewards[eid]
+
+            # Handle episode terminations
+            if dones[eid] or episode.length >= horizon:
+                done = True
+                yield CompletedRollout(episode.length, episode.total_reward)
+            else:
+                done = False
+
+            episode.batch_builder.add_values(
+                obs=episode.last_observation,
+                actions=episode.last_action_flat(),
+                rewards=rewards[eid],
+                dones=done,
+                new_obs=filtered_obs,
+                **episode.last_pi_info)
+
+            if (done and not pack) or \
+                    episode.batch_builder.count >= num_local_steps:
+                yield episode.batch_builder.build_and_reset(
+                    policy.postprocess_trajectory)
+            elif done:
+                # Make sure postprocessor never goes across episode boundaries
+                episode.batch_builder.postprocess_batch_so_far(
+                    policy.postprocess_trajectory)
+
+            if done:
+                # Handle episode termination
+                batch_builder_pool.append(episode.batch_builder)
+                del episodes[eid]
+                resetted_obs = async_vector_env.try_reset(eid)
+                if resetted_obs is None:
+                    # Reset not supported, drop this env from the ready list
+                    assert horizon == float("inf"), \
+                        "Setting episode horizon requires reset() support."
+                    ready_eids.pop()
+                    ready_obs.pop()
+                    ready_rnn_states.pop()
+                else:
+                    # Reset successful, put in the new obs as ready
+                    episode = episodes[eid]
+                    episode.last_observation = obs_filter(resetted_obs)
+                    ready_obs[-1] = episode.last_observation
+                    ready_rnn_states[-1] = episode.rnn_state
+            else:
+                episode.last_observation = filtered_obs
+
+        if not ready_eids:
+            continue  # No actions to take
+
+        # Compute action for ready envs
+        ready_rnn_state_cols = _to_column_format(ready_rnn_states)
+        actions, new_rnn_state_cols, pi_info_cols = policy.compute_actions(
+            ready_obs, ready_rnn_state_cols, is_training=True)
```

Comment:
what is `is_training` for? Couldn't find anywhere in the code where this was used in context of computing actions.

Reply:
For batch norm -- not currently used. Also, it's useful in general.